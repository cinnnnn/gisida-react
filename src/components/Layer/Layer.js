import React from 'react';
import PropTypes from 'prop-types';

const Layer = ({ mapTargetId, layer, onLayerToggle = f => f }) =>
  (<li className={`layer ${mapTargetId}`}>  
    <input
      type="checkbox"
      data-layer={layer.id}
      onChange={e => onLayerToggle(layer, e.target.checked, mapTargetId)}
    />
    <label htmlFor={layer.id} >{layer.label}</label>
  </li>);

Layer.propTypes = {
  mapTargetId: PropTypes.string.isRequired,
  layer: PropTypes.objectOf(PropTypes.any).isRequired,
  onLayerToggle: PropTypes.func.isRequired,
};

export default Layer;
