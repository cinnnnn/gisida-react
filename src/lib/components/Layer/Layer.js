import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux'
import { Actions, prepareLayer } from 'gisida'


const mapStateToProps = (state, ownProps) => {
  const MAP = state[ownProps.mapId]
  if (!!!MAP) {
    debugger
  }
  return {
    timeSeriesObj: MAP.timeseries[MAP.visibleLayerId],
    timeseries: MAP.timeseries,
    layers: MAP.layers
  }
}

export class Layer extends Component {

  onLayerToggle = (e, layer) => {
    // dispatch toggle layer action
    const mapId = this.props.mapId;
    this.props.dispatch(Actions.toggleLayer(mapId, layer.id, e.target.checked));

    // Prepare layer if layer had not been loaded
    if (!layer.loaded && !layer.isLoading) {
      prepareLayer(mapId, layer, this.props.dispatch);
    }
  }

  render() {
    const layer = this.props.layer; 
    return (
      <li className="layer">
        <input
          id={layer.id}  
          type="checkbox"
          data-layer={layer.id}
          onChange={e => this.onLayerToggle(e, layer)}
          checked={!!layer.visible}
        />
        <label htmlFor={layer.id} >{layer.label}</label>
      </li>
    );
  }
}

Layer.propTypes = {
  layer: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default connect(mapStateToProps)(Layer);
