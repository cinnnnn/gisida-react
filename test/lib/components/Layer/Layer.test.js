import React from 'react';
import Layer from '../../../../src/lib/components/Layer/Layer'
import layerObj from '../../../fixtures/sample-layer.json';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import APP from '../../../fixtures/appConfig.json';
import layer from '../../../fixtures/layers.json'

const defaultLayer = layer["Education-adolescents-15-to-18"];
defaultLayer["zoomOnToggle"] = true;
defaultLayer["visible"] = true;

const initialState = {
  "map-1": {
    timeseries: {}
  },
  APP,
  LOC: {}
}

const mockStore = configureStore()
const store = mockStore(initialState);

const mapId = 'map-1';

const componentWrapper = mount(
  <Layer
    store={store}
    layer={defaultLayer}
    mapId={mapId}
  />
);


describe('Layer', () => {

  it('component renderes correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });

  it('onLayerToggle is called when layer is checked', () => {
    global.window["maps"] = [
      {easeTo: jest.fn()}
    ];
    const mockEaseTo = jest.spyOn(window.maps[0], 'easeTo');
    componentWrapper.find('input').simulate('change');
    expect(mockEaseTo).toHaveBeenCalledTimes(1);
    delete global.window.maps;
  })

});
