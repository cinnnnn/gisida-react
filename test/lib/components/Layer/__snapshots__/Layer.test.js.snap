// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer component renderes correctly 1`] = `
<Connect(Layer)
  layer={
    Object {
      "aggregate": Object {
        "accepted-filter-values": Array [
          "all",
          "all",
        ],
        "date-by": "submission/reporting_period",
        "date-parse": Object {
          "chunk": 0,
          "split": " to ",
        },
        "extraProps": Array [
          "total_15_18_adolscent",
          "total_15_18_adolscent_girls",
          "total_15_18_adolscent_boys",
          "rc_rep/rc_boys_cal3",
          "rc_rep/rc_girls_cal3",
          "rc_rep/total3",
          "submission/location_type",
          "reporter/reporter_org",
          "Settlement",
          "submission/reporting_period",
          "New_Camp_S",
          "New_Camp_N",
          "Upazila",
        ],
        "filter": Array [
          "reporter/reporter_org",
          "Upazila",
        ],
        "filter-label": Array [
          "Partner",
          "Upazila",
        ],
        "filter-type": Array [
          "stops",
          "stops",
        ],
        "group-by": "submission/rc_loc",
        "timeseries": Object {
          "field": "period",
          "type": "cumulative",
        },
        "type": "sum",
      },
      "categories": Object {
        "breaks": "yes",
        "clusters": 7,
        "color": "RdPu",
      },
      "category": "Education",
      "credit": "Adolescents (15-18 years) enrolled in non-formal/formal education.",
      "data-parse": Object {
        "submission/rc_loc": Object {
          "key": Object {
            "CXB-017": "C￼￼￼amp 25 (Alikhali)",
            "CXB-025": "C￼￼￼amp 26 (Nayapara)",
            "CXB-032": "C￼￼￼amp 23 (Shamlapur)",
            "CXB-037": "Camp 27 (Jadimura)",
            "CXB-085": "C￼￼￼amp 22 (Unchiprang)",
            "CXB-089": "Nayapara RC",
            "CXB-108": "Camp 21 (Chakmarkul)",
            "CXB-201": "Camp 1E",
            "CXB-202": "Camp 1W",
            "CXB-203": "Camp 2E",
            "CXB-204": "Camp 2W",
            "CXB-205": "Camp 3",
            "CXB-206": "Camp 4",
            "CXB-207": "Camp 7",
            "CXB-208": "Camp 6",
            "CXB-209": "Camp 5",
            "CXB-210": "Camp 8E",
            "CXB-211": "Camp 8W",
            "CXB-212": "Camp 17",
            "CXB-213": "Camp 9",
            "CXB-214": "Camp 10",
            "CXB-215": "Camp 18",
            "CXB-216": "Camp 20",
            "CXB-217": "Camp 11",
            "CXB-218": "Camp 12",
            "CXB-219": "Camp 19",
            "CXB-220": "Camp 13 (Thangkhali-Burmapara)",
            "CXB-221": "Kutupalong RC",
            "CXB-222": "Camp 14 (Hakimapara)",
            "CXB-223": "Camp 15 (Jamtoli)",
            "CXB-224": "Camp 16 (Bagghona-Putibonia)",
            "CXB-232": "Camp 4 Extension",
            "CXB-233": "C￼￼￼amp 24 (Leda)",
            "CXB-234": "Camp 20 Extension",
          },
          "new-prop-name": "parsedUID",
        },
        "submission/reporting_period": Object {
          "key": Object {
            "t_1": "March 1 to March 20",
            "t_10": "July 11 to July 24",
            "t_11": "July 26 to August 8",
            "t_12": "August 9 to August 22",
            "t_13": "August 23 to September 5",
            "t_14": "September 6 to September 19",
            "t_15": "September 20 to October 3",
            "t_16": "October 4 to October 17",
            "t_17": "October 18 to October 31",
            "t_18": "November 1 to November 14",
            "t_19": "November 15 to November 28",
            "t_2": "March 21 to April 3",
            "t_20": "November 29 to December 12",
            "t_21": "December 13 to December 31",
            "t_3": "April 4 to April 17",
            "t_4": "April 18 to May 1",
            "t_5": "May 2 to May 15",
            "t_6": "May 16 to May 29",
            "t_7": "May 30 to June 12",
            "t_8": "June 13 to June 26",
            "t_9": "June 27 to July 10",
            "t_A17-F18": "August 2017 to February 2018",
            "t_feb": "February 1 to February 28",
            "t_jan": "January 1 to January 31",
          },
          "type": "single",
        },
      },
      "detail-view": Object {
        "UID": "SSID",
        "basic-info": Array [
          Object {
            "alt": "Location ID",
            "icon": "map-marker",
            "value": "New_Camp_S",
          },
          Object {
            "alt": "Settlement Type",
            "icon": "home",
            "value": "Settlement",
          },
          Object {
            "alt": "GPS coordinates",
            "icon": "screenshot",
            "value": Object {
              "join": ", ",
              "prop": Array [
                "longitude",
                "latitude",
              ],
            },
          },
          Object {
            "alt": "Reporter Org",
            "icon": "user",
            "value": "reporter/reporter_org",
          },
        ],
        "sub-title": Object {
          "prop": "New_Camp_N",
        },
        "title": Object {
          "prop": "Site_name",
        },
      },
      "id": "Education-adolescents-15-to-18",
      "label": "RC - Adolescents (15-18 years) enrolled in non-formal/formal education",
      "labels": Object {
        "coordinates": Array [
          "longitude",
          "latitude",
        ],
        "data": "data/180320_Outline_Rohingya_refugee_camps_sites_centroids.csv",
        "height": 30,
        "join": Array [
          "New_Camp_S",
          "submission/rc_loc",
        ],
        "label": "<b> {{total_15_18_adolscent}} </b>",
        "minZoom": 4.5,
        "width": 30,
      },
      "loaded": true,
      "popup": Object {
        "body": "<p style=text-align:left;> <b>Girls:</b> {{total_15_18_adolscent_girls}} <br> <b>Boys:</b> {{total_15_18_adolscent_boys}}<span style=float:right;>",
        "header": "parsedUID",
      },
      "property": "total_15_18_adolscent",
      "rohingyaIndicators": Object {
        "location_type": "submission/location_type",
        "rc_boys": "rc_rep/rc_boys_cal3",
        "rc_girls": "rc_rep/rc_girls_cal3",
        "rc_total": "rc_rep/total3",
      },
      "source": Object {
        "data": Array [
          315361,
          "/data/180320_Outline_Rohingya_refugee_camps_sites_centroids.csv",
        ],
        "join": Array [
          "SSID",
          "submission/rc_loc",
          "New_Camp_S",
        ],
        "layer": "180528_outline_rohingya_refug-c2ep2k",
        "relation": Object {
          "key": Array [
            "vector",
            "many",
            "one",
          ],
          "type": "one-to-many",
        },
        "type": "vector",
        "url": "mapbox://ona.5t6fsdlt",
      },
      "type": "fill",
      "visible": true,
      "zoomOnToggle": true,
    }
  }
  mapId="map-1"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Layer
    APP={
      Object {
        "accessToken": "pk.eyJ1Ijoib",
        "apiAccessToken": "ed7237b96",
        "appColor": "#0099ff",
        "appIcon": "",
        "appName": "Test App",
        "appNameDesc": "Rohingya Realtime Monitoring",
        "loaded": true,
        "mapConfig": Object {
          "center": Array [
            92.1,
            21.15,
          ],
          "container": "map-1",
          "style": "mapbox://styles/ona/cj50x94t60tlf2rn6r4axcsqz",
          "zoom": 11.5,
        },
        "password": Array [
          "unicefroh",
        ],
      }
    }
    LOC={Object {}}
    dispatch={[Function]}
    layer={
      Object {
        "aggregate": Object {
          "accepted-filter-values": Array [
            "all",
            "all",
          ],
          "date-by": "submission/reporting_period",
          "date-parse": Object {
            "chunk": 0,
            "split": " to ",
          },
          "extraProps": Array [
            "total_15_18_adolscent",
            "total_15_18_adolscent_girls",
            "total_15_18_adolscent_boys",
            "rc_rep/rc_boys_cal3",
            "rc_rep/rc_girls_cal3",
            "rc_rep/total3",
            "submission/location_type",
            "reporter/reporter_org",
            "Settlement",
            "submission/reporting_period",
            "New_Camp_S",
            "New_Camp_N",
            "Upazila",
          ],
          "filter": Array [
            "reporter/reporter_org",
            "Upazila",
          ],
          "filter-label": Array [
            "Partner",
            "Upazila",
          ],
          "filter-type": Array [
            "stops",
            "stops",
          ],
          "group-by": "submission/rc_loc",
          "timeseries": Object {
            "field": "period",
            "type": "cumulative",
          },
          "type": "sum",
        },
        "categories": Object {
          "breaks": "yes",
          "clusters": 7,
          "color": "RdPu",
        },
        "category": "Education",
        "credit": "Adolescents (15-18 years) enrolled in non-formal/formal education.",
        "data-parse": Object {
          "submission/rc_loc": Object {
            "key": Object {
              "CXB-017": "C￼￼￼amp 25 (Alikhali)",
              "CXB-025": "C￼￼￼amp 26 (Nayapara)",
              "CXB-032": "C￼￼￼amp 23 (Shamlapur)",
              "CXB-037": "Camp 27 (Jadimura)",
              "CXB-085": "C￼￼￼amp 22 (Unchiprang)",
              "CXB-089": "Nayapara RC",
              "CXB-108": "Camp 21 (Chakmarkul)",
              "CXB-201": "Camp 1E",
              "CXB-202": "Camp 1W",
              "CXB-203": "Camp 2E",
              "CXB-204": "Camp 2W",
              "CXB-205": "Camp 3",
              "CXB-206": "Camp 4",
              "CXB-207": "Camp 7",
              "CXB-208": "Camp 6",
              "CXB-209": "Camp 5",
              "CXB-210": "Camp 8E",
              "CXB-211": "Camp 8W",
              "CXB-212": "Camp 17",
              "CXB-213": "Camp 9",
              "CXB-214": "Camp 10",
              "CXB-215": "Camp 18",
              "CXB-216": "Camp 20",
              "CXB-217": "Camp 11",
              "CXB-218": "Camp 12",
              "CXB-219": "Camp 19",
              "CXB-220": "Camp 13 (Thangkhali-Burmapara)",
              "CXB-221": "Kutupalong RC",
              "CXB-222": "Camp 14 (Hakimapara)",
              "CXB-223": "Camp 15 (Jamtoli)",
              "CXB-224": "Camp 16 (Bagghona-Putibonia)",
              "CXB-232": "Camp 4 Extension",
              "CXB-233": "C￼￼￼amp 24 (Leda)",
              "CXB-234": "Camp 20 Extension",
            },
            "new-prop-name": "parsedUID",
          },
          "submission/reporting_period": Object {
            "key": Object {
              "t_1": "March 1 to March 20",
              "t_10": "July 11 to July 24",
              "t_11": "July 26 to August 8",
              "t_12": "August 9 to August 22",
              "t_13": "August 23 to September 5",
              "t_14": "September 6 to September 19",
              "t_15": "September 20 to October 3",
              "t_16": "October 4 to October 17",
              "t_17": "October 18 to October 31",
              "t_18": "November 1 to November 14",
              "t_19": "November 15 to November 28",
              "t_2": "March 21 to April 3",
              "t_20": "November 29 to December 12",
              "t_21": "December 13 to December 31",
              "t_3": "April 4 to April 17",
              "t_4": "April 18 to May 1",
              "t_5": "May 2 to May 15",
              "t_6": "May 16 to May 29",
              "t_7": "May 30 to June 12",
              "t_8": "June 13 to June 26",
              "t_9": "June 27 to July 10",
              "t_A17-F18": "August 2017 to February 2018",
              "t_feb": "February 1 to February 28",
              "t_jan": "January 1 to January 31",
            },
            "type": "single",
          },
        },
        "detail-view": Object {
          "UID": "SSID",
          "basic-info": Array [
            Object {
              "alt": "Location ID",
              "icon": "map-marker",
              "value": "New_Camp_S",
            },
            Object {
              "alt": "Settlement Type",
              "icon": "home",
              "value": "Settlement",
            },
            Object {
              "alt": "GPS coordinates",
              "icon": "screenshot",
              "value": Object {
                "join": ", ",
                "prop": Array [
                  "longitude",
                  "latitude",
                ],
              },
            },
            Object {
              "alt": "Reporter Org",
              "icon": "user",
              "value": "reporter/reporter_org",
            },
          ],
          "sub-title": Object {
            "prop": "New_Camp_N",
          },
          "title": Object {
            "prop": "Site_name",
          },
        },
        "id": "Education-adolescents-15-to-18",
        "label": "RC - Adolescents (15-18 years) enrolled in non-formal/formal education",
        "labels": Object {
          "coordinates": Array [
            "longitude",
            "latitude",
          ],
          "data": "data/180320_Outline_Rohingya_refugee_camps_sites_centroids.csv",
          "height": 30,
          "join": Array [
            "New_Camp_S",
            "submission/rc_loc",
          ],
          "label": "<b> {{total_15_18_adolscent}} </b>",
          "minZoom": 4.5,
          "width": 30,
        },
        "loaded": true,
        "popup": Object {
          "body": "<p style=text-align:left;> <b>Girls:</b> {{total_15_18_adolscent_girls}} <br> <b>Boys:</b> {{total_15_18_adolscent_boys}}<span style=float:right;>",
          "header": "parsedUID",
        },
        "property": "total_15_18_adolscent",
        "rohingyaIndicators": Object {
          "location_type": "submission/location_type",
          "rc_boys": "rc_rep/rc_boys_cal3",
          "rc_girls": "rc_rep/rc_girls_cal3",
          "rc_total": "rc_rep/total3",
        },
        "source": Object {
          "data": Array [
            315361,
            "/data/180320_Outline_Rohingya_refugee_camps_sites_centroids.csv",
          ],
          "join": Array [
            "SSID",
            "submission/rc_loc",
            "New_Camp_S",
          ],
          "layer": "180528_outline_rohingya_refug-c2ep2k",
          "relation": Object {
            "key": Array [
              "vector",
              "many",
              "one",
            ],
            "type": "one-to-many",
          },
          "type": "vector",
          "url": "mapbox://ona.5t6fsdlt",
        },
        "type": "fill",
        "visible": true,
        "zoomOnToggle": true,
      }
    }
    mapId="map-1"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    timeseries={Object {}}
  >
    <li
      className="layer"
      title=""
    >
      <input
        checked={true}
        data-layer="Education-adolescents-15-to-18"
        id="Education-adolescents-15-to-18-map-1"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="Education-adolescents-15-to-18-map-1"
      >
        RC - Adolescents (15-18 years) enrolled in non-formal/formal education
      </label>
    </li>
  </Layer>
</Connect(Layer)>
`;
